cognite/client/_api/datapoints.py

```diff
@@ -1,117 +1,969 @@
+                    rm_idx.add(i)
+                else:
+                    limits[i] -= i_part
+            not_done -= rm_idx
+        return actual_lims
+
+    @staticmethod
+    def _update_queries_is_string(res: List[DatapointsFromAPI], queries: TSQueryList) -> None:
+        is_string = {("id", r["id"]) for r in res if r["isString"]}.union(
+            ("externalId", r["externalId"]) for r in res if r["isString"]
+        )
+        for q in queries:
+            q.is_string = q.identifier.as_tuple() in is_string
+
+    @staticmethod
+    def _handle_missing_ts(
+        res: List[DatapointsFromAPI],
+        agg_queries: TSQueryList,
+        raw_queries: TSQueryList,
+    ) -> Tuple[TSQueryList, TSQueryList, Set[_SingleTSQueryBase]]:
+        missing, to_raise = set(), set()
+        not_missing = {("id", r["id"]) for r in res}.union(("externalId", r["externalId"]) for r in res)
+        for query in chain(agg_queries, raw_queries):
+            # Update _SingleTSQueryBase objects with `is_missing` status:
+            query.is_missing = query.identifier.as_tuple() not in not_missing
+            if query.is_missing:
+                missing.add(query)
+                # We might be handling multiple simultaneous top-level queries, each with a
+                # different settings for "ignore unknown":
+                if not query.ignore_unknown_ids:
+                    to_raise.add(query)
+        agg_queries = [q for q in agg_queries if not q.is_missing]
+        raw_queries = [q for q in raw_queries if not q.is_missing]
+        return agg_queries, raw_queries, to_raise
 
 
 class DatapointsAPI(APIClient):
     _RESOURCE_PATH = "/timeseries/data"
 
     def __init__(self, *args: Any, **kwargs: Any) -> None:
         super().__init__(*args, **kwargs)
-        self._DPS_LIMIT_AGG = 10000
-        self._DPS_LIMIT = 100000
-        self._POST_DPS_OBJECTS_LIMIT = 10000
-        self._RETRIEVE_LATEST_LIMIT = 100
         self.synthetic = SyntheticDatapointsAPI(
             self._config, api_version=self._api_version, cognite_client=self._cognite_client
         )
 
     def retrieve(
         self,
-        start: Union[int, str, datetime],
-        end: Union[int, str, datetime],
-        id: DatapointsIdMaybeAggregate = None,
-        external_id: DatapointsExternalIdMaybeAggregate = None,
-        aggregates: List[str] = None,
-        granularity: str = None,
-        include_outside_points: bool = None,
-        limit: int = None,
+        *,
+        id: Optional[DatapointsIdTypes] = None,
+        external_id: Optional[DatapointsExternalIdTypes] = None,
+        start: Union[int, str, datetime, None] = None,
+        end: Union[int, str, datetime, None] = None,
+        aggregates: Optional[List[str]] = None,
+        granularity: Optional[str] = None,
+        limit: Optional[int] = None,
+        include_outside_points: bool = False,
         ignore_unknown_ids: bool = False,
     ) -> Union[None, Datapoints, DatapointsList]:
```

Comment:
Let's add overloads to this method to show what the. Return type will depend on which of types included by `DatapointsIdTypes` and `DatapointsExternalIdTypes` are used, so you can't really use those type aliases then.

Reply:
There will be a looot of overloaded methods though, but I’m fine with that solution 👍 Do you see any way to avoid optional Datapoints for when “single identifier is missing and ignore=True”?

Reply:
Implemented #1051. Ended up with 3 x 8 overloaded methods 🚀